controllers/hvpa_util.go

```diff
@@ -0,0 +1,407 @@
+					// This provides hysteresis while scaling down.
+					// We can calculate new resource values based on current bucket itself then,
+					// no need to use the generic bucket list anymore - overwrite:
+					finalReplica = currentReplicas
+					break
+				}
+
+				replicaByCPU, _, _ := containerBuckets[container.Name].FindXValue(newTotalCPUReco, ResourceReplicas, ResourceCPU)
+				replicaByMem, _, _ := containerBuckets[container.Name].FindXValue(newTotalMemReco, ResourceReplicas, ResourceMemory)
+				maxReplica := int32(math.Max(float64(replicaByCPU), float64(replicaByMem)))
+
+				newPerReplicaCPU := newTotalCPUReco / int64(maxReplica)
+				newPerReplicaMem := newTotalMemReco / int64(maxReplica)
+
+				if maxReplica < currentReplicas {
+					if (currentperReplicaCPU != 0 && currentperReplicaCPU < newPerReplicaCPU) && (currentperReplicaMem != 0 && currentperReplicaMem < newPerReplicaMem) {
+						// Paradoxical Scaling - block scaling for this container
+						log.V(2).Info("Paradoxical scaling: No scaling recommended because recommendation for horizontal scale is \"scale in\", while for vertical scale is \"scale up\"", "hvpa", hvpa.Namespace+"/"+hvpa.Name)
+						blockedScalingReason[container.Name] = hvpav1alpha1.BlockingReasonParadoxicalScaling
+						break
+					}
+				}
+				if maxReplica > finalReplica {
+					finalReplica = maxReplica
+				}
+				break
+			}
+		}
+	}
+	if finalReplica == 0 {
+		// scaling for all the containers is blocked
+		finalReplica = currentReplicas
+	}
+
+	newPodSpec = podSpec.DeepCopy()
+
+	for _, reco := range vpaStatus.Recommendation.ContainerRecommendations {
+		for i := range newPodSpec.Containers {
+			container := &newPodSpec.Containers[i]
+			if container.Name == reco.ContainerName {
+				currentperReplicaCPU := container.Resources.Requests.Cpu().MilliValue()
+				currentperReplicaMem := container.Resources.Requests.Memory().MilliValue()
+
+				newTotalCPUReco := int64(math.Max(float64(reco.Target.Cpu().MilliValue()*int64(currentReplicas)), float64(container.Resources.Requests.Cpu().MilliValue()*int64(desiredReplicas))))
+				newTotalMemReco := int64(math.Max(float64(reco.Target.Memory().MilliValue()*int64(currentReplicas)), float64(container.Resources.Requests.Memory().MilliValue()*int64(desiredReplicas))))
+
+				newCPUPerReplica := newTotalCPUReco / int64(finalReplica)
+				newMemPerReplica := newTotalMemReco / int64(finalReplica)
+
+				newCPUPerReplica, newMemPerReplica = adjustForBaseUsage(newCPUPerReplica, currentperReplicaCPU, newMemPerReplica, currentperReplicaMem, finalReplica, currentReplicas, hvpa)
+
+				if finalReplica > currentReplicas {
+					if currentperReplicaCPU > newCPUPerReplica && currentperReplicaMem > newMemPerReplica {
+						// Paradoxical Scaling
+						log.V(2).Info("Paradoxical scaling: current resources recommended because recommendation for horizontal scale is \"scale out\", while for vertical scale is \"scale down\"", "hvpa", hvpa.Namespace+"/"+hvpa.Name)
```

Comment:
Should we add to `blockedScalingReason`?

Reply:
Just being consistent with the other case where we skip scaling horizontally. If we should wither add it to `blockedScaling` in both cases or neither. WDYT?

Reply:
In the other case we skip both horizontal and vertical scaling. The resources and replicas are kept same. Hence it is considered as "blocked".
In this case, we scale horizontally. We only skip scaling vertically because then it would result in paradoxical scaling. Not sure if this should be considered as "blocked". WDYT?

Reply:
I was thinking from our earlier (before the redesign) approach of updating `blockingReason` if `vpaWeight < 100` even though the recommendation is partly updated. But fair enough, it is not a strong recommendation.